src/types/widget.ts

```diff
@@ -143,62 +143,76 @@ export type RenderState = {
   [indexId: string]: IndexRenderState;
 };
 
+export type SearchBoxWidgetRenderState = WidgetRenderState<
+  {
+    query: string;
+    refine(query: string): void;
+    clear(): void;
+    isSearchStalled: boolean;
+  },
+  {
+    queryHook?(query: string, refine: (query: string) => void);
+  }
+>;
+
+export type AutocompleteWidgetRenderState = WidgetRenderState<
+  AutocompleteRendererOptions,
+  AutocompleteConnectorParams
+>;
+
+export type BreadcrumbWidgetRenderState = WidgetRenderState<
+  BreadcrumbRendererOptions,
+  BreadcrumbConnectorParams
+>;
+
+export type ClearRefinementsWidgetRenderState = WidgetRenderState<
+  ClearRefinementsRendererOptions,
+  ClearRefinementsConnectorParams
+>;
+
+export type ConfigureWidgetRenderState = WidgetRenderState<
+  ConfigureRendererOptions,
+  ConfigureConnectorParams
+>;
+
+export type CurrentRefinementsWidgetRenderState = WidgetRenderState<
+  CurrentRefinementsRendererOptions,
+  CurrentRefinementsConnectorParams
+>;
+
+export type HierarchicalMenuWidgetRenderState = WidgetRenderState<
+  {
+    items: any[];
+    refine(facetValue: any): void;
+    createURL(facetValue: any): string;
+    isShowingMore: boolean;
+    toggleShowMore(): void;
+    canToggleShowMore: boolean;
+  },
+  {
+    attributes: string[];
+    separator: string;
+    rootPath: string | null;
+    showParentLevel: boolean;
+    limit: number;
+    showMore: boolean;
+    showMoreLimit: number;
+    sortBy: any;
+    transformItems(items: any): any;
+  }
+>;
+
 export type IndexRenderState = Partial<{
-  searchBox: WidgetRenderState<
-    {
-      query: string;
-      refine(query: string): void;
-      clear(): void;
-      isSearchStalled: boolean;
-    },
-    {
-      queryHook?(query: string, refine: (query: string) => void);
-    }
-  >;
-  autocomplete: WidgetRenderState<
-    AutocompleteRendererOptions,
-    AutocompleteConnectorParams
-  >;
+  searchBox: SearchBoxWidgetRenderState;
```

Comment:
what if we write WidgetRenderState<SearchBoxRendererOptions, SearchBoxConnectorParams> directly here? Then we don't have all those types that need to be exported. The TS ones can be directly imported, the JS ones written inline (IMHO)

Having types for each individual widget here feels like it clutters this file a bit IMO

Reply:
If it definitely has to exist, I'd prefer a renderstate file (maybe a uiState file too, but out of scope of this PR, maybe make it stacked if we agree on it)

Reply:
Initially I splitted them because I had to re-use them in other places, but with the current way we don't have to. I'll merge them.

Reply:
https://github.com/algolia/instantsearch.js/pull/4518/commits/ae03fd4cda36beb0a56c88fe64592238ba465dff